presto-main/src/main/java/com/facebook/presto/execution/executor/MultilevelSplitQueue.java

```diff
@@ -0,0 +1,160 @@
+
+    public MultilevelSplitQueue()
+    {
+        this.levelMinPriority = new AtomicLong[LEVELS.length];
+        this.levelWaitingSplits = new ArrayList<>(LEVELS.length);
+        this.selectedLevelCounters = new ArrayList<>(LEVELS.length);
+
+        for (int i = 0; i < LEVELS.length; i++) {
+            levelMinPriority[i] = new AtomicLong(-1);
+            levelWaitingSplits.add(new PriorityQueue<>());
+            selectedLevelCounters.add(new CounterStat());
+        }
+
+        this.lock = new ReentrantLock();
+        this.notEmpty = lock.newCondition();
+    }
+
+    public void offer(PrioritizedSplitRunner split)
+    {
+        lock.lock();
+        split.setReady();
+        levelWaitingSplits.get(split.getLevel()).offer(split);
+        notEmpty.signal();
+        lock.unlock();
+    }
+
+    public PrioritizedSplitRunner take()
+            throws InterruptedException
+    {
+        PrioritizedSplitRunner result;
+        while (true) {
+            lock.lockInterruptibly();
+            try {
+                while ((result = takeFirstSplit()) == null) {
+                    notEmpty.await();
+                }
+
+                if (result.updateLevelPriority()) {
+                    offer(result);
+                    continue;
+                }
+
+                int selectedLevel = result.getLevel();
+                levelMinPriority[selectedLevel].set(result.getLevelPriority());
+                selectedLevelCounters.get(result.getLevel()).update(1);
+
+                return result;
+            }
+            finally {
+                lock.unlock();
+            }
+        }
+    }
+
+    /**
+     * Call only while holding lock.
+     */
+    private PrioritizedSplitRunner takeFirstSplit()
+    {
+        for (int level = 0; level < LEVELS.length; level++) {
+            PrioritizedSplitRunner split = levelWaitingSplits.get(level).poll();
+            if (split != null) {
+                return split;
+            }
+        }
+
+        return null;
+    }
+
+    public void remove(PrioritizedSplitRunner split)
+    {
+        lock.lock();
+        for (int level = 0; level < LEVELS.length; level++) {
+            levelWaitingSplits.get(level).remove(split);
+        }
+        lock.unlock();
+    }
+
+    public void removeAll(Collection<PrioritizedSplitRunner> splits)
+    {
+        lock.lock();
+        for (int level = 0; level < LEVELS.length; level++) {
+            levelWaitingSplits.get(level).removeAll(splits);
+        }
+        lock.unlock();
+    }
+
+    public long getLevelMinPriority(int level, long taskThreadUsageNanos)
+    {
+        levelMinPriority[level].compareAndSet(-1, taskThreadUsageNanos);
+        return levelMinPriority[level].get();
+    }
+
+    public int size()
+    {
+        int waitingSplits = 0;
+        for (PriorityQueue<PrioritizedSplitRunner> level : levelWaitingSplits) {
+            waitingSplits += level.size();
```

Comment:
The implementation reads an int, and I'm OK with memory-visibility effects, but I found another way.